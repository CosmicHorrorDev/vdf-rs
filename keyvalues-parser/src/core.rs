use std::{borrow::Cow, collections::BTreeMap};

pub type Key<'a> = Cow<'a, str>;
pub type Obj<'a> = BTreeMap<Key<'a>, Vec<Value<'a>>>;

#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Vdf<'a> {
    pub key: Key<'a>,
    pub value: Value<'a>,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Value<'a> {
    Str(Cow<'a, str>),
    Obj(Obj<'a>),
}

impl<'a> Value<'a> {
    pub fn is_str(&self) -> bool {
        self.get_str().is_some()
    }

    pub fn is_obj(&self) -> bool {
        self.get_obj().is_some()
    }

    pub fn get_str(&self) -> Option<&Cow<'a, str>> {
        if let Value::Str(s) = self {
            Some(s)
        } else {
            None
        }
    }

    pub fn get_obj(&self) -> Option<&Obj> {
        if let Value::Obj(obj) = self {
            Some(&obj)
        } else {
            None
        }
    }
}
