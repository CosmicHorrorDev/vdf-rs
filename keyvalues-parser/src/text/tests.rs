use std::{error::Error, fs, path::Path};

use crate::core::Vdf;

type BoxedResult<T> = Result<T, Box<dyn Error>>;

fn read_asset_file(file_name: &str) -> BoxedResult<String> {
    let val = fs::read_to_string(Path::new("tests").join("assets").join(file_name))?;
    Ok(val)
}

#[test]
fn basic() -> BoxedResult<()> {
    let vdf_text = read_asset_file("basic.vdf")?;
    let vdf = Vdf::parse(&vdf_text)?;
    insta::assert_ron_snapshot!(vdf);

    Ok(())
}

#[test]
fn app_manifest() -> BoxedResult<()> {
    let vdf_text = read_asset_file("app_manifest.vdf")?;
    let vdf = Vdf::parse(&vdf_text)?;
    insta::assert_ron_snapshot!(vdf);

    Ok(())
}
