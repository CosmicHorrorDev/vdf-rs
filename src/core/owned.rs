use std::{collections::BTreeMap, io::Read};

use crate::core::{Value, Vdf};

pub type KeyBuf = String;
pub type KeyValuesBuf = BTreeMap<KeyBuf, Vec<ValueBuf>>;

// TODO: can Vdf and VdfBuf be combined by just using `Cow<'a, str>`?
#[derive(Debug, PartialEq, Default)]
pub struct VdfBuf(pub KeyValuesBuf);

impl VdfBuf {
    // No fancy streaming or anything like that yet. This just reads the full value in and then
    // parses it
    // TODO: fix error junk here
    pub fn from_reader(mut read: impl Read) -> Result<Self, Box<dyn std::error::Error>> {
        let mut buffer = String::new();
        read.read_to_string(&mut buffer)?;

        let vdf = Vdf::parse(&buffer)?;
        Ok(vdf.into())
    }

    pub fn to_vdf(&self) -> Vdf {
        let inner = self
            .0
            .iter()
            .map(|(key, values)| {
                let key_ref = key.as_str();
                let values_ref = values.iter().map(|val| val.to_value()).collect();
                (key_ref, values_ref)
            })
            .collect();

        Vdf(inner)
    }
}

impl<'a> From<Vdf<'a>> for VdfBuf {
    fn from(vdf: Vdf) -> Self {
        vdf.into()
    }
}

impl<'a> From<&'a Vdf<'a>> for VdfBuf {
    fn from(vdf: &Vdf) -> Self {
        vdf.to_vdf_buf()
    }
}

#[derive(Debug, PartialEq)]
pub enum ValueBuf {
    Str(String),
    Obj(VdfBuf),
}

impl ValueBuf {
    pub fn to_value(&self) -> Value {
        match self {
            ValueBuf::Str(string) => Value::Str(string.as_str()),
            ValueBuf::Obj(obj) => Value::Obj(obj.to_vdf()),
        }
    }
}
