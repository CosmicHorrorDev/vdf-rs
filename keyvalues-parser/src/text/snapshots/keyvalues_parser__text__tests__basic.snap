---
source: keyvalues-parser/src/text/tests.rs
expression: vdf

---
Vdf(
  key: "Outer Key",
  value: Obj({
    "Seq Key": [
      Str("Str Val"),
      Obj({}),
    ],
  }),
)
